go-controller/pkg/node/port_claim.go

```diff
@@ -0,0 +1,181 @@
+package node
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	kapi "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog"
+)
+
+var port localPort
+
+type activeSocket interface {
+	Close() error
+}
+
+type localPort interface {
+	open(port int32, protocol kapi.Protocol, svc *kapi.Service) error
+	close(port int32, svc *kapi.Service) error
+}
+
+type portClaimWatcher struct {
+	recorder      record.EventRecorder
+	activeSockets map[int32]activeSocket
+}
+
+func newPortClaimWatcher(recorder record.EventRecorder) localPort {
+	return &portClaimWatcher{
+		recorder:      recorder,
+		activeSockets: make(map[int32]activeSocket),
+	}
+}
+
+func initPortClaimWatcher(recorder record.EventRecorder, wf *factory.WatchFactory) error {
+	port = newPortClaimWatcher(recorder)
+	_, err := wf.AddServiceHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			svc := obj.(*kapi.Service)
+			if errors := addServicePortClaim(svc); len(errors) > 0 {
+				for _, err := range errors {
+					klog.Error(err)
+				}
+			}
+		},
+		UpdateFunc: func(old, new interface{}) {
+			oldSvc := old.(*kapi.Service)
+			newSvc := new.(*kapi.Service)
+			if errors := updateServicePortClaim(oldSvc, newSvc); len(errors) > 0 {
+				for _, err := range errors {
+					klog.Error(err)
+				}
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			svc := obj.(*kapi.Service)
+			if errors := deleteServicePortClaim(svc); len(errors) > 0 {
+				for _, err := range errors {
+					klog.Error(err)
+				}
+			}
+		},
+	}, nil)
+	return err
+}
+
+func addServicePortClaim(svc *kapi.Service) []error {
+	return handleServicePortClaim(svc, true)
+}
+
+func deleteServicePortClaim(svc *kapi.Service) []error {
+	return handleServicePortClaim(svc, false)
+}
+
+func updateServicePortClaim(oldSvc, newSvc *kapi.Service) []error {
+	if reflect.DeepEqual(oldSvc.Spec.ExternalIPs, newSvc.Spec.ExternalIPs) && reflect.DeepEqual(oldSvc.Spec.Ports, newSvc.Spec.Ports) {
+		return nil
+	}
+	errors := []error{}
+	errors = append(errors, deleteServicePortClaim(oldSvc)...)
+	errors = append(errors, addServicePortClaim(newSvc)...)
+	return errors
+}
+
+func handleServicePortClaim(svc *kapi.Service, shouldOpen bool) []error {
+	errors := []error{}
+	if !util.ServiceTypeHasNodePort(svc) && len(svc.Spec.ExternalIPs) == 0 {
+		return errors
+	}
+	for _, port := range svc.Spec.Ports {
+		if util.ServiceTypeHasNodePort(svc) {
+			errors = append(errors, handlePortClaim(port.NodePort, port.Protocol, svc, shouldOpen)...)
+		}
+		if len(svc.Spec.ExternalIPs) > 0 {
+			errors = append(errors, handlePortClaim(port.Port, port.Protocol, svc, shouldOpen)...)
+		}
+	}
+	return errors
+}
+
+func handlePortClaim(portNum int32, protocol kapi.Protocol, svc *kapi.Service, shouldOpen bool) []error {
```

Comment:
i dont really like this method of passing a boolean to determine open or close and having to pass the boolean from top down. Just make an open and a close function imo